File: wifi_monitor.js
// wifi_monitor.js
const { program } = require('commander');
const axios = require('axios');
const chalk = require('chalk');
const speedtest = require('speedtest-net');
const { exec } = require('child_process');
const fs = require('fs');
const { promisify } = require('util');
const execPromise = promisify(exec);

class WiFiMonitor {
    constructor(options) {
        this.pingHost = options.ping || null;
        this.downloadSize = options.download || 0;
        this.uploadSize = options.upload || 0;
        this.count = options.count || 1;
        this.color = options.color || process.stdout.isTTY;
        this.results = [];
    }

    async measurePing(host = '8.8.8.8', count = 4) {
        try {
            const { stdout } = await execPromise(`ping -c ${count} ${host}`);
            const lines = stdout.split('\n');
            for (let line of lines) {
                if (line.includes('avg') || line.includes('rtt min/avg/max/mdev')) {
                    const match = line.match(/= ([\d.]+)\/([\d.]+)\/([\d.]+)/);
                    if (match) {
                        return parseFloat(match[2]); // avg
                    }
                }
            }
            return null;
        } catch (e) {
            return null;
        }
    }

    async measureDownload() {
        const url = 'http://ipv4.download.thinkbroadband.com/5MB.zip';
        try {
            const start = Date.now();
            const response = await axios.get(url, { responseType: 'stream' });
            let total = 0;
            return new Promise((resolve) => {
                response.data.on('data', (chunk) => { total += chunk.length; });
                response.data.on('end', () => {
                    const elapsed = (Date.now() - start) / 1000;
                    const speed = (total * 8) / (elapsed * 1_000_000);
                    resolve(speed);
                });
                response.data.on('error', () => resolve(null));
            });
        } catch (e) {
            return null;
        }
    }

    async measureUpload() {
        const url = 'https://httpbin.org/post';
        const size = this.uploadSize * 1024 * 1024;
        const data = Buffer.alloc(size, '0');
        try {
            const start = Date.now();
            await axios.post(url, data);
            const elapsed = (Date.now() - start) / 1000;
            const speed = (size * 8) / (elapsed * 1_000_000);
            return speed;
        } catch (e) {
            return null;
        }
    }

    async runTest() {
        const result = {};
        if (this.pingHost) {
            result.ping = await this.measurePing(this.pingHost);
        }
        if (this.downloadSize > 0) {
            result.download = await this.measureDownload();
        }
        if (this.uploadSize > 0) {
            result.upload = await this.measureUpload();
        }
        return result;
    }

    async run() {
        for (let i = 0; i < this.count; i++) {
            console.log(`Тест ${i+1}/${this.count}`);
            const res = await this.runTest();
            this.results.push(res);
            this.printResult(res);
            if (i < this.count - 1) await this.sleep(1000);
        }
    }

    printResult(res) {
        if (this.color) {
            if (res.ping) console.log(chalk.green(`🌐 Ping: ${res.ping.toFixed(2)} ms`));
            if (res.download) console.log(chalk.blue(`📥 Download: ${res.download.toFixed(2)} Mbps`));
            if (res.upload) console.log(chalk.cyan(`📤 Upload: ${res.upload.toFixed(2)} Mbps`));
        } else {
            if (res.ping) console.log(`Ping: ${res.ping.toFixed(2)} ms`);
            if (res.download) console.log(`Download: ${res.download.toFixed(2)} Mbps`);
            if (res.upload) console.log(`Upload: ${res.upload.toFixed(2)} Mbps`);
        }
        console.log();
    }

    sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

    saveResults(filename) {
        const ext = filename.split('.').pop().toLowerCase();
        let data;
        if (ext === 'json') {
            data = JSON.stringify(this.results, null, 2);
        } else if (ext === 'csv') {
            const header = 'ping,download,upload\n';
            const rows = this.results.map(r => `${r.ping||''},${r.download||''},${r.upload||''}`).join('\n');
            data = header + rows;
        } else {
            data = this.results.map(r => JSON.stringify(r)).join('\n');
        }
        fs.writeFileSync(filename, data);
        console.log(`Результаты сохранены в ${filename}`);
    }
}

program
    .option('-p, --ping [host]', 'Измерить пинг (по умолчанию 8.8.8.8)')
    .option('-d, --download [size]', 'Размер для загрузки в МБ', parseInt)
    .option('-u, --upload [size]', 'Размер для отдачи в МБ', parseInt)
    .option('-c, --count <number>', 'Количество тестов', parseInt, 1)
    .option('-o, --output <file>', 'Файл для сохранения')
    .option('--color', 'Принудительно включить цвет')
    .parse(process.argv);

const opts = program.opts();
if (!opts.ping && !opts.download && !opts.upload) {
    opts.ping = '8.8.8.8';
    opts.download = 5;
    opts.upload = 2;
}
const monitor = new WiFiMonitor(opts);
monitor.run().then(() => {
    if (opts.output) monitor.saveResults(opts.output);
});

File: wifi_monitor.rs
// wifi_monitor.rs
use clap::{Arg, App};
use reqwest::blocking::Client;
use serde::{Deserialize, Serialize};
use std::process::Command;
use std::fs::File;
use std::io::Write;
use std::time::{Instant, Duration};
use std::thread::sleep;
use colored::*;

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Result {
    ping: Option<f64>,
    download: Option<f64>,
    upload: Option<f64>,
}

struct Monitor {
    ping_host: Option<String>,
    download_size: usize,
    upload_size: usize,
    count: usize,
    color: bool,
    results: Vec<Result>,
}

impl Monitor {
    fn new(ping_host: Option<String>, download_size: usize, upload_size: usize, count: usize, color: bool) -> Self {
        Monitor { ping_host, download_size, upload_size, count, color, results: Vec::new() }
    }

    fn measure_ping(&self, host: &str) -> Option<f64> {
        let output = Command::new("ping").args(&["-c", "4", host]).output().ok()?;
        let s = String::from_utf8_lossy(&output.stdout);
        for line in s.lines() {
            if line.contains("avg") || line.contains("rtt min/avg/max/mdev") {
                if let Some(parts) = line.split('=').nth(1) {
                    let fields: Vec<&str> = parts.split('/').collect();
                    if fields.len() > 1 {
                        if let Ok(avg) = fields[1].parse::<f64>() {
                            return Some(avg);
                        }
                    }
                }
            }
        }
        None
    }

    fn measure_download(&self) -> Option<f64> {
        let url = "http://ipv4.download.thinkbroadband.com/5MB.zip";
        let client = Client::new();
        let start = Instant::now();
        let mut total = 0u64;
        if let Ok(mut resp) = client.get(url).send() {
            let mut buf = [0u8; 8192];
            loop {
                match resp.read(&mut buf) {
                    Ok(n) if n > 0 => { total += n as u64; }
                    Ok(_) => break,
                    Err(_) => return None,
                }
            }
            let elapsed = start.elapsed().as_secs_f64();
            let speed = (total as f64 * 8.0) / (elapsed * 1_000_000.0);
            return Some(speed);
        }
        None
    }

    fn measure_upload(&self) -> Option<f64> {
        let url = "https://httpbin.org/post";
        let size = self.upload_size * 1024 * 1024;
        let data = vec![0u8; size];
        let client = Client::new();
        let start = Instant::now();
        if client.post(url).body(data).send().is_ok() {
            let elapsed = start.elapsed().as_secs_f64();
            let speed = (size as f64 * 8.0) / (elapsed * 1_000_000.0);
            return Some(speed);
        }
        None
    }

    fn run_test(&self) -> Result {
        let mut res = Result { ping: None, download: None, upload: None };
        if let Some(ref host) = self.ping_host {
            res.ping = self.measure_ping(host);
        }
        if self.download_size > 0 {
            res.download = self.measure_download();
        }
        if self.upload_size > 0 {
            res.upload = self.measure_upload();
        }
        res
    }

    fn print_result(&self, res: &Result) {
        if self.color {
            if let Some(p) = res.ping {
                println!("{}", format!("🌐 Ping: {:.2} ms", p).green());
            }
            if let Some(d) = res.download {
                println!("{}", format!("📥 Download: {:.2} Mbps", d).blue());
            }
            if let Some(u) = res.upload {
                println!("{}", format!("📤 Upload: {:.2} Mbps", u).cyan());
            }
        } else {
            if let Some(p) = res.ping { println!("Ping: {:.2} ms", p); }
            if let Some(d) = res.download { println!("Download: {:.2} Mbps", d); }
            if let Some(u) = res.upload { println!("Upload: {:.2} Mbps", u); }
        }
        println!();
    }

    fn run(&mut self) {
        for i in 0..self.count {
            println!("Тест {}/{}", i+1, self.count);
            let res = self.run_test();
            self.results.push(res.clone());
            self.print_result(&res);
            if i < self.count - 1 {
                sleep(Duration::from_secs(1));
            }
        }
    }

    fn save(&self, filename: &str) -> std::io::Result<()> {
        let ext = filename.split('.').last().unwrap_or("txt").to_lowercase();
        let data: Vec<u8> = match ext.as_str() {
            "json" => serde_json::to_string_pretty(&self.results).unwrap().into_bytes(),
            "csv" => {
                let mut s = "ping,download,upload\n".to_string();
                for r in &self.results {
                    s.push_str(&format!("{},{},{}\n",
                        r.ping.map_or("".to_string(), |v| v.to_string()),
                        r.download.map_or("".to_string(), |v| v.to_string()),
                        r.upload.map_or("".to_string(), |v| v.to_string())
                    ));
                }
                s.into_bytes()
            }
            _ => format!("{:?}", self.results).into_bytes()
        };
        let mut file = File::create(filename)?;
        file.write_all(&data)?;
        Ok(())
    }
}

fn main() {
    let matches = App::new("WiFi Monitor")
        .arg(Arg::with_name("ping").long("ping").takes_value(true).help("Хост для пинга"))
        .arg(Arg::with_name("download").long("download").takes_value(true).help("Размер загрузки в МБ"))
        .arg(Arg::with_name("upload").long("upload").takes_value(true).help("Размер отдачи в МБ"))
        .arg(Arg::with_name("count").long("count").takes_value(true).default_value("1").help("Количество тестов"))
        .arg(Arg::with_name("output").long("output").takes_value(true).help("Файл для сохранения"))
        .arg(Arg::with_name("color").long("color").help("Принудительный цвет"))
        .get_matches();

    let ping_host = matches.value_of("ping").map(|s| s.to_string());
    let download_size = matches.value_of("download").and_then(|s| s.parse::<usize>().ok()).unwrap_or(0);
    let upload_size = matches.value_of("upload").and_then(|s| s.parse::<usize>().ok()).unwrap_or(0);
    let count = matches.value_of("count").and_then(|s| s.parse::<usize>().ok()).unwrap_or(1);
    let output = matches.value_of("output").map(|s| s.to_string());
    let color = matches.is_present("color") || atty::is(atty::Stream::Stdout);

    if ping_host.is_none() && download_size == 0 && upload_size == 0 {
        // default all
        let mut m = Monitor::new(Some("8.8.8.8".to_string()), 5, 2, count, color);
        m.run();
        if let Some(f) = output {
            if let Err(e) = m.save(&f) {
                eprintln!("Ошибка сохранения: {}", e);
            } else {
                println!("Результаты сохранены в {}", f);
            }
        }
    } else {
        let mut m = Monitor::new(ping_host, download_size, upload_size, count, color);
        m.run();
        if let Some(f) = output {
            if let Err(e) = m.save(&f) {
                eprintln!("Ошибка сохранения: {}", e);
            } else {
                println!("Результаты сохранены в {}", f);
            }
        }
    }
}
